app/src/main/java/com/dimagi/biometric/fragments/BaseMatchFragment.java

```diff
@@ -0,0 +1,107 @@
+import androidx.activity.result.ActivityResultLauncher;
+import androidx.activity.result.contract.ActivityResultContracts;
+import androidx.appcompat.app.AlertDialog;
+import androidx.core.content.ContextCompat;
+import androidx.fragment.app.Fragment;
+
+import android.util.Log;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.Button;
+import android.widget.TextView;
+import android.widget.Toast;
+
+import com.dimagi.biometric.R;
+
+/**
+ * A base class to contain common functionality for both the face and finger match fragments.
+ */
+public abstract class BaseMatchFragment extends Fragment {
+    private static final String TAG = "BIOMETRIC";
+
+    private Toast permissionToast;
+
+    protected abstract void handleStartCapture();
+    protected abstract void handleCancelCapture();
+
+    public BaseMatchFragment() {
+        // Required empty public constructor
+    }
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
+        View view = inflater.inflate(R.layout.fragment_main_menu, container, false);
+
+        Button startButton = view.findViewById(R.id.start_capture_button);
+        startButton.setOnClickListener(v -> {
+            if (ContextCompat.checkSelfPermission(
+                    requireContext(), Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED) {
+                handleStartCapture();
+            } else if (shouldShowRequestPermissionRationale(Manifest.permission.CAMERA)) {
+                showPermissionAlertDialog();
+            } else {
+                requestPermissionLauncher.launch(Manifest.permission.CAMERA);
+            }
+        });
+        Button cancelButton = view.findViewById(R.id.cancel_capture_button);
+        cancelButton.setOnClickListener(v -> handleCancelCapture());
+
+        return view;
+    }
+
+    private final ActivityResultLauncher<String> requestPermissionLauncher =
+            registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
+        if (isGranted) {
+            handleStartCapture();
+        } else {
+            showPermissionToast();
+        }
+    });
+
+    private void showPermissionAlertDialog() {
+        AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(requireContext());
+        alertDialogBuilder.setMessage(getText(R.string.camera_permission_rationale));
+        alertDialogBuilder.setPositiveButton(getText(R.string.confirm), new DialogInterface.OnClickListener() {
+            @Override
+            public void onClick(DialogInterface dialog, int which) {
+                requestPermissionLauncher.launch(Manifest.permission.CAMERA);
+            }
+        });
+
+        alertDialogBuilder.setNegativeButton(getText(R.string.cancel), new DialogInterface.OnClickListener() {
+            @Override
+            public void onClick(DialogInterface dialog, int which) {
```

Comment:
Currently what happens is that the user is unable to start the capture process until the permission has been given. Without the permission, all the user can do is click the "Cancel" button to cancel the biometric workflow. @avazirna Do you think a different action should be taken when the permission is denied?

Reply:
@zandre-eng  Right now, in case the user doesn't grant the permission in the first attempt, a `toast` is presented to them informing that the permission is required and the process ends there. Only when they hit `Start` again they are able to see the permission rationale, we should show the rationale in that first attempt. Additionally, we need to be more specific about the consequences of not granting the permission, what features are still available to the user without this permission?

Reply:
I have reworked the implementation for the permission checking ([b77ac30](https://github.com/dimagi/biometric-android/pull/1/commits/b77ac306dda86310c35755598edd7c3c894dd1b6)) so that it now gives a dialog instead of a toast to prompt the user to enable the required permission in their settings if it has been denied twice.